ndm-exporter/collector/seachest.go

```diff
@@ -0,0 +1,207 @@
+// correspnding to each blockdevice
+type SeachestMetricData struct {
+	SeachestIdentifier *seachest.Identifier
+	TempInfo           blockdevice.TemperatureInformation
+}
+
+// NewSeachestMetricCollector creates a new instance of SeachestCollector which
+// implements Collector interface
+func NewSeachestMetricCollector(c kubernetes.Client) prometheus.Collector {
+	klog.V(2).Infof("Seachest Metric Collector initialized")
+	sc := &SeachestCollector{
+		Client:  c,
+		metrics: smartmetrics.NewMetrics(SeachestCollectorNamespace),
+	}
+	sc.metrics.WithBlockDeviceCurrentTemperature().
+		WithBlockDeviceCurrentTemperatureValid().
+		WithRejectRequest().
+		WithErrorRequest()
+	return sc
+}
+
+// Describe is the implementation of Describe in prometheus.Collector
+func (sc *SeachestCollector) Describe(ch chan<- *prometheus.Desc) {
+	for _, col := range sc.metrics.Collectors() {
+		col.Describe(ch)
+	}
+}
+
+// Collect is the implementation of Collect in prometheus.Collector
+func (sc *SeachestCollector) Collect(ch chan<- prometheus.Metric) {
+	klog.V(4).Info("Starting to collect smartmetrics metrics for a request")
+
+	sc.Lock()
+	if sc.requestInProgress {
+		klog.V(4).Info("Another request already in progress.")
+		sc.metrics.IncRejectRequestCounter()
+		sc.Unlock()
+		return
+	}
+
+	sc.requestInProgress = true
+	sc.Unlock()
+
+	// once a request is processed, set the progress flag to false
+	defer sc.setRequestProgressToFalse()
+
+	klog.V(4).Info("Setting client for this request.")
+
+	// set the client each time
+	if err := sc.Client.InitClient(); err != nil {
+		klog.Errorf("error setting client. %v", err)
+		sc.metrics.IncErrorRequestCounter()
+		sc.collectErrors(ch)
+		return
+	}
+
+	// get list of blockdevices from etcd
+	blockDevices, err := sc.Client.ListBlockDevice()
+	if err != nil {
+		sc.metrics.IncErrorRequestCounter()
+		sc.collectErrors(ch)
+		return
+	}
+
+	klog.V(4).Info("Blockdevices fetched from etcd")
+
+	err = GetMetricData(blockDevices)
+	if err != nil {
+		sc.metrics.IncErrorRequestCounter()
+		sc.collectErrors(ch)
+		return
+	}
+
+	klog.V(4).Infof("metrics data obtained from seachest library")
+
+	sc.SetMetricData(blockDevices)
+
+	klog.V(4).Info("Prometheus metrics is set and initializing collection.")
+
+	// collect each metric
+	for _, col := range sc.metrics.Collectors() {
+		col.Collect(ch)
+	}
+}
+
+// setRequestProgressToFalse is used to set the progress flag, when a request is
+// processed or errored
+func (sc *SeachestCollector) setRequestProgressToFalse() {
+	sc.Lock()
+	sc.requestInProgress = false
+	sc.Unlock()
+}
+
+// collectErrors collects only the error metrics and set it on the channel
+func (sc *SeachestCollector) collectErrors(ch chan<- prometheus.Metric) {
+	for _, col := range sc.metrics.ErrorCollectors() {
+		col.Collect(ch)
+	}
+}
+
+// GetMetricData gets the seachest metrics for each blockdevice and fills it in the blockdevice struct
```

Comment:
Should this method be exposed outside of the package? Similar check needs to be performed to other method to set the visibility as appropriate.

Reply:
Done. methods are now not available outside the package